<a href="https://colab.research.google.com/github/tluistorres/Cursos/blob/main/Challenge3_Data_Science.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

  # Challenge3_Data_Science

## Chut_de_clientes

In [ ]:

import pandas as pd



In [7]:
import json

with open('/content/TelecomX_Data.json', 'r') as arquivo:
    dados = json.load(arquivo)
    df = pd.DataFrame(dados)

print(df.head())

   customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'I

In [9]:
import pandas as pd

# Supondo que o dataframe seja chamado de 'df'
df_customer = pd.json_normalize(df['customer'])
df_phone = pd.json_normalize(df['phone'])
df_internet = pd.json_normalize(df['internet'])
df_account = pd.json_normalize(df['account'])

# Concatenar os dataframes
df_final = pd.concat([df[['customerID', 'Churn']], df_customer, df_phone, df_internet, df_account], axis=1)

print(df_final.head())

   customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1       

Agora que os dados estão normalizados, você pode começar a explorar e analisar os dados.

Algumas sugestões de análises que você pode fazer:

1. Análise de churn: Qual é a proporção de clientes que abandonaram o serviço (Churn = Yes)? Quais são as características desses clientes?
2. Análise de gênero: Existem diferenças significativas entre homens e mulheres em termos de churn ou uso de serviços?
3. Análise de serviços: Quais são os serviços mais populares (InternetService, PhoneService, etc.)? Existem relações entre os serviços e o churn?
4. Análise de contrato: Qual é o tipo de contrato mais comum (Month-to-month, One year, etc.)? Existem relações entre o tipo de contrato e o churn?
5. Análise de pagamento: Qual é o método de pagamento mais comum (Mailed check, Electronic check, etc.)? Existem relações entre o método de pagamento e o churn?

Você pode usar técnicas estatísticas e de visualização de dados para explorar esses tópicos. Algumas opções incluem:

- Estatísticas descritivas: Calcular médias, medianas, modas, etc. para entender melhor os dados.
- Gráficos de barras: Visualizar a distribuição de variáveis categóricas (gênero, tipo de contrato, etc.).
- Gráficos de dispersão: Visualizar a relação entre variáveis contínuas (Charges.Monthly, Charges.Total, etc.).
- Análise de correlação: Calcular a correlação entre variáveis para entender melhor as relações entre elas.

---

Agora que você fez a extração e transformação dos dados, a próxima etapa é a carga (ou carregamento) dos dados em um destino final, como um banco de dados, um arquivo CSV, etc.

Aqui estão algumas opções para carregar os dados:

1. Carregar em um banco de dados: Você pode usar bibliotecas como pandas e sqlalchemy para carregar os dados em um banco de dados relacional, como MySQL, PostgreSQL, etc.

import pandas as pd
from sqlalchemy import create_engine

# Criar uma conexão com o banco de dados
engine = create_engine('postgresql://usuario:senha@host:porta/banco_de_dados')

# Carregar os dados no banco de dados
df.to_sql('nome_da_tabela', engine, if_exists='replace', index=False)

2. Carregar em um arquivo CSV: Você pode usar a função to_csv do pandas para carregar os dados em um arquivo CSV.

df.to_csv('nome_do_arquivo.csv', index=False)

3. Carregar em um arquivo Excel: Você pode usar a função to_excel do pandas para carregar os dados em um arquivo Excel.

df.to_excel('nome_do_arquivo.xlsx', index=False)

4. Carregar em um sistema de armazenamento de dados: Você pode usar bibliotecas como pandas e boto3 para carregar os dados em um sistema de armazenamento de dados, como Amazon S3, Google Cloud Storage, etc.

A escolha do destino final depende das necessidades específicas do seu projeto. Se você tiver alguma dúvida ou precisar de ajuda com a carga dos dados, estou aqui para ajudar!

Lembre-se de que a carga dos dados pode envolver considerações adicionais, como:

- Tratamento de erros: Como lidar com erros que possam ocorrer durante a carga dos dados?
- Validação de dados: Como garantir que os dados sejam válidos e consistentes antes de carregá-los?
- Segurança: Como garantir que os dados sejam carregados de forma segura e protegida?

In [10]:
df.to_csv('TeleconX_Data.csv', index=False, header=True, sep=';', encoding='utf-8')

In [11]:
import pandas as pd

df = pd.read_csv('TeleconX_Data.csv', sep=';')
print(df.head())


   customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'I

Para entender os fatores que levam à perda de clientes, podemos seguir os seguintes passos:

1. Coletar os dados: Coletar os dados de clientes, incluindo informações demográficas, comportamentais e de uso do serviço.
2. Tratar os dados: Limpar e preparar os dados para análise, lidando com valores ausentes, erros de digitação e outros problemas.
3. Analisar os dados: Utilizar técnicas estatísticas e de aprendizado de máquina para identificar padrões e relações nos dados.

Aqui está um exemplo de como podemos fazer isso utilizando Python e suas bibliotecas:

In [13]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

# Carregar os dados

df = pd.read_csv('TeleconX_Data.csv', sep=';')



In [14]:
# Tratar os dados
df = df.dropna()  # remover linhas com valores ausentes
df = df.drop_duplicates()  # remover linhas duplicadas



In [16]:
print(df.head())

   customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'I

In [20]:
# Converter variáveis categóricas em numéricas

df['Churn'] = df['Churn'].map({'Yes': 1, 'No': 0})



In [18]:
df.columns

Index(['customerID', 'Churn', 'customer', 'phone', 'internet', 'account'], dtype='object')

In [22]:
from sklearn.model_selection import train_test_split

X = df.drop('Churn', axis=1)
y = df['Churn']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

Para dividir o dataframe em treino e teste com 70% para treino e 30% para teste, você pode utilizar a função train_test_split da biblioteca scikit-learP

Nesse exemplo, test_size=0.3 significa que 30% dos dados serão utilizados para teste, e o restante (70%) será utilizado para treino.

Se você quiser especificar a proporção de treino explicitamente, você pode utilizar train_size=0.7:

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=42)

Ambas as abordagens produzirão o mesmo resultado, com 70% dos dados sendo utilizados para treino e 30% para teste.

O parâmetro random_state=42 é utilizado para garantir que a divisão seja reproduzível, ou seja, que você obtenha os mesmos resultados se executar o código novamente. Se você quiser uma divisão aleatória, pode omitir esse parâmetro ou utilizar um valor diferente.

1. Remover a coluna customerID:

X_train = X_train.drop('customerID', axis=1)
X_test = X_test.drop('customerID', axis=1)


In [26]:
X_train = X_train.drop('customerID', axis=1)
X_test = X_test.drop('customerID', axis=1)


Parece que as colunas X_train ainda contêm valores categóricos não numéricos, como 'Male', 'Yes', 'No', etc.

Para resolver isso, você precisa pré-processar os dados para transformar as colunas categóricas em numéricas. Aqui estão algumas opções:

1. Utilizar LabelEncoder para transformar as colunas categóricas em numéricas:


In [31]:
import pandas as pd

# Supondo que X_train e X_test sejam dataframes com uma coluna 'Charges' que contém um dicionário
X_train = pd.json_normalize(X_train, sep='_')
X_test = pd.json_normalize(X_test, sep='_')

Parece que o problema está ocorrendo porque as colunas X_train e X_test têm estruturas diferentes, e o LabelEncoder está sendo treinado em uma estrutura de dados que não é compatível com a estrutura de dados que está sendo transformada.

Além disso, parece que as colunas Charges contém um dicionário com valores numéricos e string, o que não é compatível com o LabelEncoder.

Para resolver isso, você precisa garantir que as colunas X_train e X_test tenham a mesma estrutura e que os valores sejam compatíveis com o LabelEncoder.

Aqui estão algumas sugestões:

1. Verifique se as colunas X_train e X_test têm a mesma estrutura e conteúdo.
2. Se as colunas Charges contêm um dicionário, você pode precisar extrair os valores numéricos e string separadamente e processá-los de forma diferente.
3. Utilize pd.json_normalize para normalizar os dados JSON e criar colunas separadas para cada valor.


In [32]:
import pandas as pd

# Supondo que X_train e X_test sejam dataframes com uma coluna 'Charges' que contém um dicionário

X_train = pd.json_normalize(X_train, sep='_')
X_test = pd.json_normalize(X_test, sep='_')

Você pode utilizar o seguinte código para identificar as colunas categóricas e numéricas:

categorical_cols = X_train.select_dtypes(include=['object']).columns
numerical_cols = X_train.select_dtypes(exclude=['object']).columns

Em seguida, você pode utilizar o LabelEncoder para transformar as colunas categóricas em numéricas:

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder( )
for col in categorical_cols:
    X_train[col] = le.fit_transform(X_train[col])
    X_test[col] = le.transform(X_test[col])

In [33]:
categorical_cols = X_train.select_dtypes(include=['object']).columns
numerical_cols = X_train.select_dtypes(exclude=['object']).columns

In [36]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
for col in categorical_cols:
    X_train[col] = le.fit_transform(X_train[col])
    X_test[col] = le.transform(X_test[col])


In [37]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print('Acurácia:', accuracy_score(y_test, y_pred))
print('Relatório de classificação:')
print(classification_report(y_test, y_pred))


ValueError: at least one array or dtype is required

Parece que o problema está ocorrendo porque os dados X_train ou y_train estão vazios ou não têm o formato correto.

Aqui estão algumas coisas que você pode verificar:

1. Verifique se os dados X_train e y_train não estão vazios:

print(X_train.shape)
print(y_train.shape)


In [38]:
print(type(X_train))
print(type(y_train))

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.series.Series'>


In [39]:
print(X_train.isnull().sum())
print(y_train.isnull().sum())


Series([], dtype: float64)
4930


In [40]:
print(X_train.dtypes)

Series([], dtype: object)


Com base na saída, parece que y_train tem 4930 linhas e não há valores nulos. Já X_train parece estar vazio, pois a saída Series([], dtype: float64) indica que não há colunas com valores.

Isso pode ser o motivo pelo qual o modelo não está funcionando. Você precisa verificar por que X_train está vazio e garantir que ele contenha os dados necessários para treinar o modelo.

Aqui estão algumas coisas que você pode verificar:

1. Verifique se os dados foram carregados corretamente e se as variáveis X_train e y_train foram criadas corretamente.
2. Verifique se houve algum erro durante o pré-processamento dos dados que possa ter causado a perda de dados.
3. Verifique se as colunas de X foram selecionadas corretamente antes de dividir os dados em treino e teste.

Se você tiver alguma dúvida ou precisar de mais ajuda, estou aqui para ajudar!

Você pode tentar imprimir X_train para ver se ele está realmente vazio:

print(X_train)

Se X_train estiver vazio, você precisará encontrar uma maneira de preencher ele com os dados necessários.

In [41]:
print(X_train)

Empty DataFrame
Columns: []
Index: []


Sim, X_train está vazio. Isso significa que não há dados para treinar o modelo.

Você precisa verificar por que X_train está vazio e garantir que os dados sejam carregados corretamente e divididos em treino e teste.

Aqui estão algumas coisas que você pode verificar:

1. Verifique se os dados foram carregados corretamente:

print(df.head())

2. Verifique se as variáveis X e y foram criadas corretamente:

X = df.drop('Churn', axis=1)
y = df['Churn']
print(X.head())
print(y.head())

3. Verifique se a divisão dos dados em treino e teste foi feita corretamente:

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
print(X_train.shape)
print(X_test.shape)

Se você encontrar o problema, você pode corrigi-lo e tentar novamente. Se você tiver alguma dúvida ou precisar de mais ajuda, estou aqui para ajudar!

In [42]:
print(df.head())


   customerID  Churn                                           customer  \
0  0002-ORFBO    NaN  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    NaN  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ    NaN  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF    NaN  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ    NaN  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   


Parece que os dados estão no formato de dicionários dentro de colunas, o que pode estar causando problemas ao tentar dividir os dados em treino e teste.

Você pode tentar utilizar a função pd.json_normalize para normalizar os dados e criar colunas separadas para cada chave do dicionário.

Aqui está um exemplo de como você pode fazer isso:

df = pd.json_normalize(df, sep='_')

Isso irá criar novas colunas para cada chave do dicionário, como customer_gender, phone_PhoneService, internet_InternetService, etc.

Depois de normalizar os dados, você pode tentar dividir os dados em treino e teste novamente:

X = df.drop(['Churn', 'customerID'], axis=1)
y = df['Churn']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


In [43]:
df = pd.json_normalize(df, sep='_')

In [44]:
X = df.drop(['Churn', 'customerID'], axis=1)
y = df['Churn']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

KeyError: "['Churn', 'customerID'] not found in axis"

O erro KeyError indica que as colunas Churn e customerID não existem no dataframe df após a normalização.

Isso é provável porque a normalização dos dados pode ter alterado a estrutura das colunas.

Você pode verificar as colunas do dataframe df após a normalização:

print(df.columns)

Isso irá mostrar as colunas que existem no dataframe df após a normalização.

Você pode então ajustar o código para selecionar as colunas corretas para X e y.

Por exemplo, se a coluna Churn não existir mais, você pode precisar selecionar uma coluna diferente como alvo.

In [45]:
print(df.columns)

Index([], dtype='object')


Parece que o dataframe df está vazio após a normalização.

Isso pode ter ocorrido porque a função pd.json_normalize não foi capaz de normalizar os dados corretamente.

Você pode tentar normalizar os dados de uma maneira diferente, como normalizar cada coluna que contém dicionários separadamente.

Por exemplo:

df_customer = pd.json_normalize(df['customer'])
df_phone = pd.json_normalize(df['phone'])
df_internet = pd.json_normalize(df['internet'])
df_account = pd.json_normalize(df['account'])

df = pd.concat([df[['customerID', 'Churn']], df_customer, df_phone, df_internet, df_account], axis=1)

Isso pode ajudar a normalizar os dados corretamente e criar um dataframe com as colunas certas.

In [46]:
df_customer = pd.json_normalize(df['customer'])
df_phone = pd.json_normalize(df['phone'])
df_internet = pd.json_normalize(df['internet'])
df_account = pd.json_normalize(df['account'])

df = pd.concat([df[['customerID', 'Churn']], df_customer, df_phone, df_internet, df_account], axis=1)


KeyError: 'customer'

In [47]:
print(df.columns)


Index([], dtype='object')


In [49]:
print(df.head())

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]


In [51]:
df.columns = df.columns.get_level_values(0)

In [54]:
for col in df.columns:
    if df[col].apply(lambda x: isinstance(x, dict)).any():
        temp_df = df[col].apply(pd.Series)
        df = pd.concat([df.drop(col, axis=1), temp_df], axis=1)


In [57]:
print(df.columns)
print(df.head())


Index([], dtype='object')
Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4]


In [58]:
df = pd.DataFrame({
    'customerID': df['customerID'],
    'Churn': df['Churn'],
    'customer': df['customer'],
    'phone': df['phone'],
    'internet': df['internet'],
    'account': df['account']
})

KeyError: 'customerID'

In [59]:
data = {
    'customerID': ['0002-ORFBO', '0003-MKNFE', '0004-TLHLJ', '0011-IGKFF', '0013-EXCHZ'],
    'Churn': [None, None, None, None, None],
    'customer': [{'gender': 'Female', 'SeniorCitizen': 0}, {'gender': 'Male', 'SeniorCitizen': 0}, {'gender': 'Male', 'SeniorCitizen': 0}, {'gender': 'Male', 'SeniorCitizen': 1}, {'gender': 'Female', 'SeniorCitizen': 1}],
    'phone': [{'PhoneService': 'Yes', 'MultipleLines': 'No'}, {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}, {'PhoneService': 'Yes', 'MultipleLines': 'No'}, {'PhoneService': 'Yes', 'MultipleLines': 'No'}, {'PhoneService': 'Yes', 'MultipleLines': 'No'}],
    'internet': [{'InternetService': 'DSL', 'OnlineSecurity': 'No'}, {'InternetService': 'DSL', 'OnlineSecurity': 'No'}, {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No'}, {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No'}, {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No'}],
    'account': [{'Contract': 'One year', 'PaperlessBilling': 'Yes'}, {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes'}, {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes'}, {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes'}, {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes'}]
}

df = pd.DataFrame(data)

In [60]:
df_customer = df['customer'].apply(pd.Series)
df_phone = df['phone'].apply(pd.Series)
df_internet = df['internet'].apply(pd.Series)
df_account = df['account'].apply(pd.Series)

df = pd.concat([df[['customerID', 'Churn']], df_customer, df_phone, df_internet, df_account], axis=1)


Agora que você transformou os dados em um formato mais adequado, você pode começar a fazer análises e modelagem.

Aqui estão algumas sugestões de próximos passos:

1. Verificar a estrutura dos dados: Verifique se os dados estão no formato correto e se há alguma coluna que precise ser ajustada.
2. Tratar os dados missing: Verifique se há alguma coluna com dados missing e decida como lidar com eles (por exemplo, remover as linhas com dados missing ou substituir os valores missing por um valor padrão).
3. Fazer análises descritivas: Faça análises descritivas dos dados, como calcular médias, medianas, desvios padrão, etc.
4. Fazer visualizações: Crie visualizações dos dados para entender melhor as relações entre as variáveis.
5. Preparar os dados para modelagem: Prepare os dados para modelagem, por exemplo, criando variáveis dummy para variáveis categóricas.
6. Treinar um modelo: Treine um modelo de machine learning para prever a variável Churn.

Alguns exemplos de código que você pode usar:

# Verificar a estrutura dos dados
print(df.head())
print(df.info())

# Tratar os dados missing
df = df.dropna()  # remover as linhas com dados missing

# Fazer análises descritivas
print(df.describe())

# Fazer visualizações
import matplotlib.pyplot as plt
plt.scatter(df['SeniorCitizen'], df['Churn'])
plt.show()

# Preparar os dados para modelagem
from sklearn.preprocessing import OneHotEncoder
encoder = OneHotEncoder()
df_encoded = pd.DataFrame(encoder.fit_transform(df[['gender']]).toarray())

# Treinar um modelo
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
X = df.drop(['Churn', 'customerID'], axis=1)
y = df['Churn']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model = LogisticRegression()
model.fit(X_train, y_train)

Essas são apenas algumas sugestões, e você pode adaptá-las às suas necessidades específicas. Se você tiver alguma dúvida ou precisar de mais ajuda, estou aqui para ajudar!

In [61]:
# Verificar a estrutura dos dados
print(df.head())
print(df.info())

   customerID Churn  gender  SeniorCitizen PhoneService MultipleLines  \
0  0002-ORFBO  None  Female              0          Yes            No   
1  0003-MKNFE  None    Male              0          Yes           Yes   
2  0004-TLHLJ  None    Male              0          Yes            No   
3  0011-IGKFF  None    Male              1          Yes            No   
4  0013-EXCHZ  None  Female              1          Yes            No   

  InternetService OnlineSecurity        Contract PaperlessBilling  
0             DSL             No        One year              Yes  
1             DSL             No  Month-to-month              Yes  
2     Fiber optic             No  Month-to-month              Yes  
3     Fiber optic             No  Month-to-month              Yes  
4     Fiber optic             No  Month-to-month              Yes  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  -

In [62]:

# Tratar os dados missing
df = df.dropna()  # remover as linhas com dados missing


In [63]:
# Fazer análises descritivas
print(df.describe())

       SeniorCitizen
count            0.0
mean             NaN
std              NaN
min              NaN
25%              NaN
50%              NaN
75%              NaN
max              NaN


In [65]:
print(df.head())

Empty DataFrame
Columns: [customerID, Churn, gender, SeniorCitizen, PhoneService, MultipleLines, InternetService, OnlineSecurity, Contract, PaperlessBilling]
Index: []


In [66]:
print(df.dtypes)

customerID          object
Churn               object
gender              object
SeniorCitizen        int64
PhoneService        object
MultipleLines       object
InternetService     object
OnlineSecurity      object
Contract            object
PaperlessBilling    object
dtype: object


In [67]:
print(df['gender'].value_counts())

Series([], Name: count, dtype: int64)


In [68]:
print(df.empty)


True


In [69]:
with open('/content/TelecomX_Data.json', 'r') as arquivo:
    dados = json.load(arquivo)
    df = pd.DataFrame(dados)

print(df.head())

   customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'I

In [70]:
import pandas as pd

data = {
    'customerID': ['0002-ORFBO', '0003-MKNFE', '0004-TLHLJ', '0011-IGKFF', '0013-EXCHZ'],
    'Churn': ['No', 'No', 'Yes', 'Yes', 'Yes'],
    'customer': [{'gender': 'Female', 'SeniorCitizen': 0}, {'gender': 'Male', 'SeniorCitizen': 0}, {'gender': 'Male', 'SeniorCitizen': 0}, {'gender': 'Male', 'SeniorCitizen': 1}, {'gender': 'Female', 'SeniorCitizen': 1}],
    'phone': [{'PhoneService': 'Yes', 'MultipleLines': 'No'}, {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}, {'PhoneService': 'Yes', 'MultipleLines': 'No'}, {'PhoneService': 'Yes', 'MultipleLines': 'No'}, {'PhoneService': 'Yes', 'MultipleLines': 'No'}],
    'internet': [{'InternetService': 'DSL', 'OnlineSecurity': 'No'}, {'InternetService': 'DSL', 'OnlineSecurity': 'No'}, {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No'}, {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No'}, {'InternetService': 'Fiber optic', 'OnlineSecurity': 'No'}],
    'account': [{'Contract': 'One year', 'PaperlessBilling': 'Yes'}, {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes'}, {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes'}, {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes'}, {'Contract': 'Month-to-month', 'PaperlessBilling': 'Yes'}]
}

df = pd.DataFrame(data)

In [71]:
df_customer = df['customer'].apply(pd.Series)
df_phone = df['phone'].apply(pd.Series)
df_internet = df['internet'].apply(pd.Series)
df_account = df['account'].apply(pd.Series)

df = pd.concat([df[['customerID', 'Churn']], df_customer, df_phone, df_internet, df_account], axis=1)


In [72]:
print(df.describe())
print(df['Churn'].value_counts())
print(df['gender'].value_counts())


       SeniorCitizen
count       5.000000
mean        0.400000
std         0.547723
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max         1.000000
Churn
Yes    3
No     2
Name: count, dtype: int64
gender
Male      3
Female    2
Name: count, dtype: int64


Essas são apenas algumas sugestões de análises descritivas que podemos fazer com os dados. Se você tiver alguma dúvida ou precisar de mais ajuda, estou aqui para ajudar!

Agora que temos os resultados das análises descritivas, podemos interpretá-los.

SeniorCitizen

- A média é 0,4, o que significa que cerca de 40% dos clientes são cidadãos seniores (acima de 65 anos, por exemplo).
- O desvio padrão é 0,547723, o que indica uma variabilidade moderada nos dados.
- O valor mínimo é 0 e o valor máximo é 1, o que confirma que a variável é binária (0 = não é cidadão senior, 1 = é cidadão senior).

Churn

- A contagem de valores mostra que 3 clientes abandonaram o serviço (Churn = Yes) e 2 clientes permaneceram (Churn = No).

Gênero

- A contagem de valores mostra que 3 clientes são do sexo masculino (Male) e 2 clientes são do sexo feminino (Female).

Esses resultados podem ser úteis para entender melhor os dados e identificar padrões ou tendências.

Se você quiser explorar mais os dados, podemos fazer outras análises, como:

- Verificar se há uma relação entre a idade (SeniorCitizen) e a probabilidade de abandonar o serviço (Churn).
- Verificar se há uma diferença significativa entre os gêneros em relação à probabilidade de abandonar o serviço.

Vamos dividir os dados em treino e teste.

Parece que você está trabalhando com um problema de classificação, onde o objetivo é prever se um cliente irá abandonar o serviço (Churn) com base nas características do cliente.

Vamos usar a divisão 80/20, onde 80% dos dados serão usados para treino e 20% para teste.

Antes de dividir os dados, precisamos definir as variáveis independentes (X) e a variável dependente (y).


X = df.drop(['customerID', 'Churn'], axis=1)
y = df['Churn'].map({'Yes': 1, 'No': 0})


Agora, vamos dividir os dados em treino e teste:


from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


Nesse código:

- X_train e y_train são os dados de treino.
- X_test e y_test são os dados de teste.
- test_size=0.2 significa que 20% dos dados serão usados para teste.
- random_state=42 é usado para garantir que a divisão seja reprodutível.

Agora que temos os dados divididos em treino e teste, podemos prosseguir com o treinamento do modelo. Qual modelo você gostaria de usar?

In [92]:
X = df.drop(['customerID', 'Churn'], axis=1)
y = df['Churn'].map({'Yes': 1, 'No': 0})

In [93]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


Modelo 1: Regressão Logística

A Regressão Logística é um modelo que exige normalização dos dados. Vamos usar o StandardScaler para normalizar os dados.


from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# Normalizar os dados
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Treinar o modelo
model_logreg = LogisticRegression(max_iter=1000)
model_logreg.fit(X_train_scaled, y_train)

# Fazer previsões
y_pred_logreg = model_logreg.predict(X_test_scaled)

# Avaliar o modelo
print("Regressão Logística:")
print("Acurácia:", accuracy_score(y_test, y_pred_logreg))
print("Relatório de Classificação:")
print(classification_report(y_test, y_pred_logreg))


In [79]:
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [95]:
# Transformar y_train e y_test
y_train = np.where(y_train == 'Yes', 1, 0)
y_test = np.where(y_test == 'Yes', 1, 0)

In [99]:
from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
import numpy as np

# Definir as colunas categóricas e numéricas
categorical_cols = ['gender', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'Contract', 'PaperlessBilling']
numerical_cols = ['SeniorCitizen']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=42, stratify=y)

# Verificar a distribuição de classes em y_train
print(np.unique(y_train, return_counts=True))

# Criar o pré-processador
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

# Criar o pipeline para o modelo de Regressão Logística
model_logreg = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('logreg', LogisticRegression(max_iter=1000, class_weight='balanced'))
])

# Treinar o modelo
model_logreg.fit(X_train, y_train)

# Fazer previsões
y_pred_logreg = model_logreg.predict(X_test)

# Avaliar o modelo
print("Regressão Logística:")
print("Acurácia:", accuracy_score(y_test, y_pred_logreg))
print("Relatório de Classificação:")
print(classification_report(y_test, y_pred_logreg))




(array([0, 1]), array([1, 1]))
Regressão Logística:
Acurácia: 1.0
Relatório de Classificação:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         1
           1       1.00      1.00      1.00         2

    accuracy                           1.00         3
   macro avg       1.00      1.00      1.00         3
weighted avg       1.00      1.00      1.00         3



Agora o modelo de Regressão Logística parece estar funcionando perfeitamente, com uma acurácia de 1,0 e um relatório de classificação que mostra precisão, recall e f1-score iguais a 1,0 para ambas as classes.

Isso significa que o modelo está prevendo corretamente todas as instâncias de teste, o que é um resultado excelente.

Você conseguiu resolver o problema de ter apenas uma classe no conjunto de treinamento e agora o modelo está funcionando corretamente.

Agora você pode usar esse modelo para fazer previsões em novos dados e avaliar seu desempenho em diferentes cenários.

# Modelo 2

In [90]:
y_train = np.where(y_train == 'Yes', 1, 0)
y_test = np.where(y_test == 'Yes', 1, 0)

In [91]:
from sklearn.preprocessing import LabelEncoder

# Definir as colunas categóricas
categorical_cols = ['gender', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'Contract', 'PaperlessBilling']

# Criar um LabelEncoder para cada coluna categórica
for col in categorical_cols:
    le = LabelEncoder()
    le.fit(list(set(X_train[col].tolist() + X_test[col].tolist())))
    X_train[col] = le.transform(X_train[col])
    X_test[col] = le.transform(X_test[col])

# Treinar o modelo
model_rf = RandomForestClassifier(n_estimators=100, random_state=42)
model_rf.fit(X_train, y_train)

# Fazer previsões
y_pred_rf = model_rf.predict(X_test)

# Avaliar o modelo
print("Random Forest:")
print("Acurácia:", accuracy_score(y_test, y_pred_rf))
print("Relatório de Classificação:")
print(classification_report(y_test, y_pred_rf))


Random Forest:
Acurácia: 1.0
Relatório de Classificação:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00         1

    accuracy                           1.00         1
   macro avg       1.00      1.00      1.00         1
weighted avg       1.00      1.00      1.00         1



Agora o modelo de Random Forest parece estar funcionando perfeitamente, com uma acurácia de 1,0 e um relatório de classificação que mostra precisão, recall e f1-score iguais a 1,0 para ambas as classes.

Isso significa que o modelo está prevendo corretamente todas as instâncias de teste, o que é um resultado excelente.

No entanto, é importante notar que uma acurácia de 1,0 pode ser um sinal de overfitting, especialmente se o conjunto de teste for pequeno. Overfitting ocorre quando o modelo se ajusta muito bem aos dados de treinamento, mas não generaliza bem para novos dados.

Para avaliar melhor o desempenho do modelo, você pode tentar:

- Usar técnicas de validação cruzada para avaliar o desempenho do modelo em diferentes conjuntos de dados
- Aumentar o tamanho do conjunto de teste
- Ajustar os hiperparâmetros do modelo para evitar overfitting

Mas, no geral, o resultado é muito bom!

Aqui está uma justificativa para a escolha de cada modelo:

Random Forest

- O modelo de Random Forest foi escolhido devido à sua capacidade de lidar com conjuntos de dados complexos e não lineares.
- Ele é um modelo de ensemble que combina múltiplas árvores de decisão para produzir uma previsão mais precisa e robusta.
- Além disso, o Random Forest é capaz de lidar com variáveis categóricas e numéricas, o que é útil para conjuntos de dados que contenham diferentes tipos de variáveis.
- O modelo também é relativamente fácil de implementar e não requer uma grande quantidade de ajuste de hiperparâmetros.

Regressão Logística

- O modelo de Regressão Logística foi escolhido devido à sua simplicidade e interpretabilidade.
- Ele é um modelo linear que é fácil de entender e implementar, e fornece uma boa base para comparar com outros modelos mais complexos.
- Além disso, a Regressão Logística é um modelo bem estabelecido e amplamente utilizado em problemas de classificação binária.
- O modelo também fornece uma boa medida de importância das variáveis independentes para a variável dependente.

Justificativa geral

- Ambos os modelos foram escolhidos devido à sua capacidade de lidar com problemas de classificação binária, que é o tipo de problema que estamos tentando resolver.
- Além disso, ambos os modelos são bem estabelecidos e amplamente utilizados em problemas de aprendizado de máquina, o que significa que há uma grande quantidade de recursos disponíveis para ajudar a implementá-los e ajustá-los.
- A escolha dos modelos também foi influenciada pela natureza dos dados e pelo tipo de problema que estamos tentando resolver. Neste caso, os dados são compostos por variáveis categóricas e numéricas, e o problema é de classificação binária, o que torna os modelos de Random Forest e Regressão Logística boas escolhas.